File: ZigZagString/ZigZagString.py

class Solution(object):

    def __init__(self):
        self.s = input()
        self.n = int(input())
        self.ans = [[] for _ in range(self.n)]

        self.findSolution()

    def findSolution(self):

        row = 0
        down = True

        for i in range(len(self.s)):

            self.ans[row] += self.s[i]

            if row == 0:
                down = True

            if row == self.n-1:
                down = False

            if down:
                row += 1

            else:
                row -= 1

        res = []
        for i in self.ans:
            res += i

        self.ans = "".join(res)

    def __del__(self):
        print(self.ans)


def main():
    sol = Solution()


if __name__ == '__main__':
    main()

File: ZigZagString/ZigZagString.cpp
#include<bits/stdc++.h>
#define int long long
using namespace std;

class Solution {

    public:
        int n;
        string s, answer;

        void findSolution();
        Solution() {
            cin>>n;
            s = "PAYPALISHIRING";
            findSolution();
        }

};

void Solution::findSolution() {

    int row = 0;
    bool down = true;

    vector<char> ans[n];

    for(auto c: s) {

        ans[row].emplace_back(c);

        if (row == 0) down = true;
        if (row == n-1) down = false;

        if (down) row++;
        else row--;
    }

    for (int i=0;i<n;i++) {
        for(int j=0;j<ans[i].size();++j) {
            // cout << ans[i][j];
            answer+=ans[i][j];
        }
    }

    cout<<answer;
}


int32_t main() {

    int t;
    cin>>t;

    while(t--) {
        Solution *sol = new Solution();
        delete sol;
    }
}